core/src/main/scala/org/apache/spark/deploy/history/EventLogFileCompactor.scala

```diff
@@ -0,0 +1,235 @@
+      }
+    }
+  }
+
+  private def findFilesToCompact(
+      eventLogFiles: Seq[FileStatus]): (Seq[FileStatus], Seq[FileStatus]) = {
+    val lastCompactedFileIdx = eventLogFiles.lastIndexWhere { fs =>
+      EventLogFileWriter.isCompacted(fs.getPath)
+    }
+    val files = eventLogFiles.drop(lastCompactedFileIdx)
+
+    if (files.length > maxFilesToRetain) {
+      (files.dropRight(maxFilesToRetain), files.takeRight(maxFilesToRetain))
+    } else {
+      (Seq.empty, files)
+    }
+  }
+}
+
+class FilteredEventLogFileRewriter(
+    sparkConf: SparkConf,
+    hadoopConf: Configuration,
+    eventLogFiles: Seq[FileStatus],
+    fs: FileSystem,
+    filters: Seq[EventFilter]) extends Logging {
+
+  require(eventLogFiles.nonEmpty)
+
+  private val targetEventLogFilePath = eventLogFiles.last.getPath
+  private val logWriter: CompactedEventLogFileWriter = new CompactedEventLogFileWriter(
+    targetEventLogFilePath, "dummy", None, targetEventLogFilePath.getParent.toUri,
+    sparkConf, hadoopConf)
+
+  def logPath: String = logWriter.logPath
+
+  def start(): Unit = {
+    logWriter.start()
+  }
+
+  def stop(): Unit = {
+    logWriter.stop()
+  }
+
+  def rewrite(): Unit = {
+    eventLogFiles.foreach(rewriteFile)
+  }
+
+  private def rewriteFile(fileStatus: FileStatus): Unit = {
+    Utils.tryWithResource(EventLogFileReader.openEventLog(fileStatus.getPath, fs)) { in =>
+      val lines = Source.fromInputStream(in)(Codec.UTF8).getLines()
+
+      var currentLine: String = null
+      var lineNumber: Int = 0
+
+      try {
+        val lineEntries = lines.zipWithIndex
+
+        while (lineEntries.hasNext) {
+          try {
+            val entry = lineEntries.next()
+
+            currentLine = entry._1
+            lineNumber = entry._2 + 1
+
+            val event = JsonProtocol.sparkEventFromJson(parse(currentLine))
+            if (checkFilters(event)) {
+              logWriter.writeLine(currentLine)
+            }
+          } catch {
+            // ignore any exception occurred from unidentified json
+            // just skip handling and write the line
+            case _: ClassNotFoundException => logWriter.writeLine(currentLine)
+            case _: UnrecognizedPropertyException => logWriter.writeLine(currentLine)
+          }
+        }
+        true
+      } catch {
+        case e: Exception =>
+          logError(s"Exception parsing Spark event log: ${fileStatus.getPath.getName}", e)
+          logError(s"Malformed line #$lineNumber: $currentLine\n")
+          throw e
+      }
+    }
+  }
+
+  private def checkFilters(event: SparkListenerEvent): Boolean = {
+    val results = filters.flatMap(filter => applyFilter(filter, event))
+    results.isEmpty || results.forall(_ == true)
+  }
+
+  private def applyFilter(filter: EventFilter, event: SparkListenerEvent): Option[Boolean] = {
+    event match {
+      case event: SparkListenerStageSubmitted => filter.filterStageSubmitted(event)
```

Comment:
Only slightly touched the filtering part (and no formed view on this) but I'm always afraid a bit when I see switch case with event types. It's extremely easy to add +-1 to the list and it's always questionable when new event type comes where other programmer has no idea that this part must be modified to reach proper implementation. Will comes back to this when fully processed...

Reply:
Yeah you're right. This class actually extended SparkListener to guarantee same coverage but I realized I shouldn't rewrite the json but keep the origin json so this is the result. 

So the listener should be aware of origin line along with event instance to write origin line after passing filter. Maybe we can have one, not sure it's somewhat beauty, but may be able to address inconsistency between this pattern matching and SparkListener. Let me try to address this.

Reply:
Btw even the case we can miss adding filter for newer event types, but we can make sure these lines are not dropped, then that should be OK.

Reply:
Please consider my last comment here; once this is providing safety I'd just keep it as it is. Injecting SparkListener into this logic makes thing being complicated.

Reply:
Please check the UT of FilteredEventLogFileRewriterSuite out, it covers None & None.